tests/pipeline/pipeline.cpp

```diff
@@ -0,0 +1,356 @@
+#include <livre/core/pipeline/Filter.h>
+#include <livre/core/pipeline/Pipeline.h>
+#include <livre/core/pipeline/SimpleExecutor.h>
+#include <livre/core/pipeline/Workers.h>
+#include <livre/core/pipeline/FutureMap.h>
+#include <livre/core/pipeline/PromiseMap.h>
+#include <livre/core/pipeline/Future.h>
+#include <livre/core/pipeline/Promise.h>
+
+#include <boost/test/unit_test.hpp>
+
+namespace ut = boost::unit_test;
+
+const uint32_t defaultMeaningOfLife = 42;
+const uint32_t defaultThanksForAllTheFish = 51;
+const uint32_t addMoreFish = 10;
+
+struct InputData
+{
+    explicit InputData( uint32_t value = defaultMeaningOfLife )
+        : meaningOfLife( value )
+    {}
+
+    uint32_t meaningOfLife;
+};
+
+struct OutputData
+{
+    explicit OutputData( uint32_t value = defaultThanksForAllTheFish )
+        : thanksForAllTheFish( value )
+    {}
+
+    uint32_t thanksForAllTheFish;
+};
+
+class TestFilter : public livre::Filter
+{
+    void execute( const livre::FutureMap& input, livre::PromiseMap& output ) const final
+    {
+        const std::vector< InputData >& results = input.get< InputData >( "TestInputData" );
+        OutputData outputData;
+
+        for( const auto& data: results )
+        {
+             const InputData& inputData = data;
+             outputData.thanksForAllTheFish += inputData.meaningOfLife + addMoreFish;
+        }
+
+        output.set( "TestOutputData", outputData );
+    }
+
+    livre::DataInfos getInputDataInfos() const final
+    {
+        return
+        {
+            { "TestInputData", livre::getType< InputData >( )}
+        };
+    }
+
+    livre::DataInfos getOutputDataInfos() const final
+    {
+        return
+        {
+            { "TestOutputData", livre::getType< OutputData >( )}
+        };
+    }
+};
+
+class ConvertFilter : public livre::Filter
+{
+    void execute( const livre::FutureMap& input, livre::PromiseMap& output ) const final
+    {
+        const std::vector< OutputData >& results = input.get< OutputData >( "ConvertInputData" );
+
+        InputData inputData;
+        for( const auto& data: results )
+        {
+             inputData.meaningOfLife = data.thanksForAllTheFish + addMoreFish;
+        }
+
+        output.set( "ConvertOutputData", inputData );
+    }
+
+    livre::DataInfos getInputDataInfos() const final
+    {
+        return
+        {
+            { "ConvertInputData", livre::getType< OutputData >( )}
+        };
+
+    }
+
+    livre::DataInfos getOutputDataInfos() const final
+    {
+        return
+        {
+            { "ConvertOutputData", livre::getType< InputData >( )}
+        };
+    }
+};
+
+bool check_error( const std::runtime_error& ) { return true; }
+
+BOOST_AUTO_TEST_CASE( testFilterNoInput )
+{
+    livre::PipeFilterT< TestFilter > pipeFilter( "Producer" );
+
+    // Execute will fail because there are no inputs where data is retrieved
+    BOOST_CHECK_EXCEPTION( pipeFilter.execute(), std::runtime_error, check_error );
```

Comment:
Since this is a programming error, doesn't it make sense to throw logic_error instead of runtime_error?